noaa.go

```diff
@@ -33,90 +33,147 @@ type PointsResponse struct {
 }
 
 // StationsResponse holds the JSON values from /points/<lat,lon>/stations
 type StationsResponse struct {
 	Stations []string `json:"observationStations"`
 }
 
+// ForecastElevaion holds the JSON values for a forecast response's elevation.
+type ForecastElevation struct {
+	Value float64 `json:"value"`
+	Units string  `json:"unitCode"`
+}
+
+type ForecastHourlyElevation struct {
+	Value          float64 `json:"value"`
+	Max            float64 `json:"maxValue"`
+	Min            float64 `json:"minValue"`
+	UnitCode       string  `json:"unitCode"`
+	QualityControl string  `json:"qualityControl"`
+}
+
+// Period holds the JSON values for a period within a forecast response's periods.
+type ForecastResponsePeriod struct {
+	ID              int32   `json:"number"`
+	Name            string  `json:"name"`
+	StartTime       string  `json:"startTime"`
+	EndTime         string  `json:"endTime"`
+	IsDaytime       bool    `json:"isDaytime"`
+	Temperature     float64 `json:"temperature"`
+	TemperatureUnit string  `json:"temperatureUnit"`
+	WindSpeed       string  `json:"windSpeed"`
+	WindDirection   string  `json:"windDirection"`
+	Icon            string  `json:"icon"`
+	Summary         string  `json:"shortForecast"`
+	Details         string  `json:"detailedForecast"`
+}
+
+type ForecastResponsePeriodHourly struct {
+	ID               int32   `json:"number"`
+	Name             string  `json:"name"`
+	StartTime        string  `json:"startTime"`
+	EndTime          string  `json:"endTime"`
+	IsDaytime        bool    `json:"isDaytime"`
+	TemperatureTrend *string `json:"temperatureTrend"`
+	WindDirection    string  `json:"windDirection"`
+	Summary          string  `json:"shortForecast"`
+	Icon             string  `json:"icon,omitempty"`
+	Details          string  `json:"detailedForecast"`
+}
+
 // ForecastResponse holds the JSON values from /gridpoints/<cwa>/<x,y>/forecast"
 type ForecastResponse struct {
 	// capture data from the forecast
-	Updated   string `json:"updated"`
-	Units     string `json:"units"`
-	Elevation struct {
-		Value float64 `json:"value"`
-		Units string  `json:"unitCode"`
-	} `json:"elevation"`
-	Periods []struct {
-		ID              int32   `json:"number"`
-		Name            string  `json:"name"`
-		StartTime       string  `json:"startTime"`
-		EndTime         string  `json:"endTime"`
-		IsDaytime       bool    `json:"isDaytime"`
-		Temperature     float64 `json:"temperature"`
-		TemperatureUnit string  `json:"temperatureUnit"`
-		WindSpeed       string  `json:"windSpeed"`
-		WindDirection   string  `json:"windDirection"`
-		Summary         string  `json:"shortForecast"`
-		Details         string  `json:"detailedForecast"`
-	} `json:"periods"`
-	Point *PointsResponse
+	Updated   string                    `json:"updated"`
+	Units     string                    `json:"units"`
+	Elevation *ForecastElevation        `json:"elevation"`
+	Periods   []*ForecastResponsePeriod `json:"periods"`
```

Comment:
I'm OK with breaking out the types into separate structs, and I know some of this already existed in the module, but I'd actually like to get rid of most of the pointers in this to help reduce memory allocations. A slice of pointers in this case could probably just be a slice of ForecastResponsePeriod structs. This would be the same for the other single-item pointers as well. I have not started that work yet, but you are welcome to give it a go in this PR if you want to clean it up along with your other changes.

Reply:
Not a problem.